csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -253,14 +255,99 @@ public TBuilder Resize(int length)
 
             public TBuilder Swap(int i, int j)
             {
-                // TODO: Implement
-                throw new NotImplementedException();
+                // Retrieve byte arrays for elements i and j
+                var iBytes = GetValueBytes(i);
+                var jBytes = GetValueBytes(j);
+
+                Set(i, jBytes);
+                Set(j, iBytes);
+
+                ValidityBuffer.Swap(i, j);
+                return Instance;
             }
 
+
             public TBuilder Set(int index, byte value)
             {
-                // TODO: Implement
-                throw new NotImplementedException();
+                int currentStart = ValueOffsets.Span[index];
+                int currentEnd = ValueOffsets.Span[index + 1];
+                int newEnd = currentStart + 1;
+
+                int currentLength = currentEnd - currentStart;
+                int newSize = 1;
+                int diffSize = currentLength - newSize;
+                if (currentLength != 1)
+                {
+                    ValueBuffer.Span.Slice(currentStart, currentLength).Clear();
+                    ValueBuffer.Span[currentStart] = value;
+                    ValueBuffer.Span.Slice(currentEnd, Length - currentEnd)
+                        .CopyTo(ValueBuffer.Span.Slice(newEnd, Length - newEnd));
+                    ValueBuffer.Span.Slice(Length - diffSize, diffSize).Clear();
+                    AdjustOffsets(index, currentLength - 1);
+                }
+
+                ValueBuffer.Span[index] = value;
+                ValidityBuffer.Set(index);
+                return Instance;
+            }
+
+            public TBuilder Set(int index, ReadOnlySpan<byte> bytes)
+            {
+                int startOffset = ValueOffsets.Span[index];
+                int newLength = bytes.Length;
+                bool isNull = bytes.IsEmpty;
```

Comment:
I don't think this code uses an empty span to mean null; an empty span is just a value of length zero.